File: app/src/main/java/co/ostorlab/insecure_app/bugs/calls/SQLiteDatabaseCall.java
package co.ostorlab.insecure_app.bugs.calls;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import co.ostorlab.insecure_app.BugRule;

public class SQLiteDatabaseCall extends BugRule {

    @Override
    public void run() throws Exception {

        MySQLiteOpenHelper mySQLiteOpenHelper = new MySQLiteOpenHelper(this.getContext());
        SQLiteDatabase db = mySQLiteOpenHelper.getWritableDatabase();
        mySQLiteOpenHelper.createTable();
        String insert_query = "INSERT INTO accounts(name, amount) VALUES(?, ?)";
        db.execSQL(insert_query, new Object[]{"Jack", 3000});
        mySQLiteOpenHelper.dropTable();
        db.close();

    }

    @Override
    public String getDescription() {
        return "The application uses SQLiteOpenHelper";
    }
}


class MySQLiteOpenHelper extends SQLiteOpenHelper {

    // database name
    private static final String DATABASE_NAME = "bank";

    // table name
    private static final String TABLE_NAME = "accounts";

    // database version
    private static final int DATABASE_VERSION = 1;

    // names of the table columns
    private static final String UID = "_id";
    private static final String NAME = "name";
    private static final String AMOUNT = "amount";

    // sql to create table
    private static final String CREATE_TABLE = "CREATE TABLE " + TABLE_NAME
            + " ("
            + UID + " INTEGER PRIMARY KEY AUTOINCREMENT, "
            + NAME + " VARCHAR(255), "
            + AMOUNT + " REAL"
            + ");";

    // sql to drop table
    private static final String DROP_TABLE = "DROP TABLE IF EXISTS " + TABLE_NAME;

    // custom constructor
    public MySQLiteOpenHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL(CREATE_TABLE);
    }

    public void createTable(){
        this.getWritableDatabase().execSQL(CREATE_TABLE);
    }

    public void dropTable(){
        this.getWritableDatabase().execSQL(DROP_TABLE);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        db.execSQL(DROP_TABLE);
        onCreate(db);
    }
}

File: ostorlab_insecure_flutter_app/lib/bugs/sqlite_database_call.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:ostorlab_insecure_flutter_app/bug_rule.dart';
import 'package:path/path.dart';
import 'package:sqflite/sqflite.dart';

class SQLiteDatabaseCall extends BugRule {
  @override
  Future<void> run() async {
    SqfliteHelper sqfliteHelper = SqfliteHelper();
    Database db = await sqfliteHelper.getWritableDatabase();

    // get user input for name and amount from a text field
    TextEditingController nameController = TextEditingController(text: "ostorlab_user");
    TextEditingController amountController = TextEditingController(text: "0");

    String name = nameController.text;
    double amount = double.tryParse(amountController.text) ?? 0.0;

    // insert data into database
    String insertQuery = "INSERT INTO ${SqfliteHelper.TABLE_NAME}(${SqfliteHelper.NAME}, ${SqfliteHelper.AMOUNT}) VALUES(?, ?)";
    await db.rawInsert(insertQuery, [name, amount]);

    sqfliteHelper.dropTable();
    db.close();
  }


  @override
  String get description => 'The application uses sqflite';
}

class SqfliteHelper {
  // database name
  static const String DATABASE_NAME = 'bank';

  // table name
  static const String TABLE_NAME = 'accounts';

  // database version
  static const int DATABASE_VERSION = 1;

  // names of the table columns
  static const String UID = '_id';
  static const String NAME = 'name';
  static const String AMOUNT = 'amount';

  // sql to create table
  static const String CREATE_TABLE = 'CREATE TABLE $TABLE_NAME '
      '($UID INTEGER PRIMARY KEY AUTOINCREMENT, '
      '$NAME VARCHAR(255), '
      '$AMOUNT REAL'
      ');';

  // sql to drop table
  static const String DROP_TABLE = 'DROP TABLE IF EXISTS $TABLE_NAME';

  // custom constructor
  SqfliteHelper();

  FutureOr<Database> getWritableDatabase() async {
    var databasesPath = await getDatabasesPath();
    var path = join(databasesPath, DATABASE_NAME);

    return await openDatabase(path, version: DATABASE_VERSION,
        onCreate: (Database db, int version) async {
      await db.execute(CREATE_TABLE);
    });
  }

  void createTable() async {
    Database db = await getWritableDatabase();
    await db.execute(CREATE_TABLE);
    db.close();
  }

  void dropTable() async {
    Database db = await getWritableDatabase();
    await db.execute(DROP_TABLE);
    db.close();
  }
}
